daml-assistant/daml-project-config/DA/Daml/Project/Config.hs

```diff
@@ -63,6 +76,85 @@ readConfig name path = do
+interpolateEnvVariables :: [(String, String)] -> T.Text -> Either String T.Text
+interpolateEnvVariables env str = do
+  let matchPositions :: [[SubMatch]] = transformSubmatch $ getAllTextMatches (str =~ ("(^|[^\\\\])(\\\\*)\\${([^}]+)}" :: String))
+      -- First 2 matches aren't needed (whole match + start of line check)
+      --   TDFA doesn't support non capturing groups, which is why that match exists in the first place
+      replaceWithPrefix (s, start) [_, _, escapeSlashes, name] = do
+        -- Divide by 2 and floor to get the number of `\` that should be in the final string
+        let finalSlashesText = T.replicate (smLength escapeSlashes `div` 2) "\\"
+            prefix = textSub start (smStart escapeSlashes) str
+        if smLength escapeSlashes `mod` 2 == 1 -- Odd number of backslashes, thiss expression is escaped
+          then pure
+                ( s <> prefix
+                    <> finalSlashesText
+                    <> textSub (smEnd escapeSlashes) (smEnd name + 1) str
+                , smEnd name + 1
+                )
+          else do
+            let envVarName = T.unpack $ T.replace "." "_" $ smText name
+                mEnvVar = lookup envVarName env
+            envVarValue <- maybeToEither ("Couldn't find environment variable " <> envVarName <> " in value " <> T.unpack str) mEnvVar
+            pure 
+              ( s <> prefix <> finalSlashesText <> T.pack envVarValue
+              , smEnd name + 1
+              )
+      -- Impossible case to appease the warnings
+      replaceWithPrefix _ _ = Left "Impossible incorrect regex submatches"
+  (replacedPrefix, prefixEnd) <- foldlM replaceWithPrefix ("" :: Text, 0) matchPositions
+  pure $ replacedPrefix <> textSub prefixEnd (T.length str) str
+
+-- | (internal) Helper function for defining 'readXConfig' functions.
+-- Throws a ConfigError if reading or parsing fails.
+-- Also performs environment variable interpolation on all string fields in the form of ${ENV_VAR}.
+-- TODO: ensure everything that reads the daml.yaml uses this, e.g. assistant for daml.yaml, multi-package for dependencies, etc.
+readConfigWithEnv :: Y.FromJSON b => Text -> FilePath -> IO b
+readConfigWithEnv name path = do
+  configE <- runExceptT $ do
+    (configValue :: Y.Value) <- ExceptT $ Y.decodeFileEither path
+    env <- liftIO getEnvironment
+    configValueTransformed <- 
+      except $ flip transformM configValue $ \case
+        Y.String str -> bimap Y.AesonException Y.String $ interpolateEnvVariables env str
+        v -> pure v
+    case fromJSON configValueTransformed of
+      Success x -> pure x
+      Error str -> throwE $ Y.AesonException str
+  fromRightM (throwIO . ConfigFileInvalid name) configE
+
+-- Returns false if the file doesn't exist or isn't valid yaml.
+-- Does not attempt to replace values, so will not fail if the environment variables do not exist.
+configExistsWithEnv :: FilePath -> IO Bool
```

Comment:
Doesn't this function also return false if the file doesn't contain any env var interpolations? It would be good to document that in the comment above.

Reply:
Yes, I had assumed this was implied by the name but I see it very much isn't. I'll rename and update the comment